[BugFix] incorrect column merging in SelectOperator::pull_chunk

xhumanoid opened the pull request:
`_pre_output_chunk` and `_curr_chunk` can have different columns order inside

if types on `append` is matched, final data will be incorrect (merge data from different columns)

if types mismatched, BE will crash

## Why I'm doing:

## What I'm doing:

Fixes #56588

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

stdpain commented:
Is it possible to provide a minimum reproduce case?

xhumanoid commented:
it's difficult to provide reproducible example, because issue happens only in large queries
but what i found

AnalyticSourceOperator is a first operator in pipeline
in pipeline we have AnalyticSourceOperator (with window function) and after next SelectOperator
 
I added very simple code in `AnalyticSourceOperator::pull_chunk` to iterate over `_slot_id_to_index`

```
for (auto slot : chunk->get_slot_id_to_index_map()) {
    std::cout << slot.first << "   " << slot.second << std::endl;
}
```

and can see 2 types of output 

```
954   2
814   5
629   0
955   7
653   1
620   3
813   4
848   6
```
and
```
954   2
814   5
848   6
955   7
653   1
620   3
629   0
813   4
```

technically both of chunks fully identical and if we order by column position it will be fully the same

```
629   0
653   1
954   2
620   3
813   4
814   5
848   6
955   7
```
different only in order of iteration over `_slot_id_to_index`
and this order is a main issue

after `AnalyticSourceOperator::pull_chunk` we will have `SelectOperator::push_chunk` where we have code

https://github.com/StarRocks/starrocks/blob/main/be/src/exec/pipeline/select_operator.cpp#L90-L92
```
for (const auto& [slot_id, _] : chunk->get_slot_id_to_index_map()) {
    _curr_chunk->append_column(chunk->get_column_by_slot_id(slot_id), slot_id);
}
```

different order of `slot_id` from iterator -> different order of columns in final chunk

later when later we try to merge this chunks in `SelectOperator::pull_chunk` 

`_slot_id_to_index` is a `phmap::flat_hash_map<SlotId, size_t, StdHash<SlotId>>`
where in comments we have description

```
// An `phmap::flat_hash_map<K, V>` is an unordered associative container which
// has been optimized for both speed and memory footprint in most common use
// cases. Its interface is similar to that of `std::unordered_map<K, V>` with
// the following notable differences:
```

**unordered** associative container
so we can't rely on order elements in iterator, to generate new vector with columns 

fix can be implemented similar in one of the way:
1. as proposed in current PR - on merge don't rely on order of column and always select them by slot_id
2. maybe more correct to do fix in `SelectOperator::push_chunk` and do any sorts by `slot_id`, in this case we will have some guarantees on column order in produced chunks

stdpain commented:
@xhumanoid  Can you provide a query profile? I will confirm this further.

stdpain commented:
Some operators of StarRocks will depend on the order of Chunk's _columns. But this order is not determined by the iteration order of _slot_id_to_index, but by append_column.

```
void Chunk::append_column(ColumnPtr column, SlotId slot_id) {
    DCHECK(!_slot_id_to_index.contains(slot_id)) << "slot_id:" + std::to_string(slot_id) << std::endl;
    if (UNLIKELY(_slot_id_to_index.contains(slot_id))) {
        throw std::runtime_error(fmt::format("slot_id {} already exists", slot_id));
    }
    _slot_id_to_index[slot_id] = _columns.size();
    _columns.emplace_back(std::move(column));
    check_or_die();
}
```

So the different iteration order of _slot_id_to_index is not a problem.
Consider adding the following Assert codes
```
    if (slotmap.empty()) {
        slotmap = chunk->get_slot_id_to_index_map();
    }
    CHECK(slotmap == chunk->get_slot_id_to_index_map());
```

Method to get query profile: https://docs.starrocks.io/docs/sql-reference/sql-functions/utility-functions/get_query_profile/

The root cause may be in other operators.

xhumanoid commented:
issue started to trigger after

https://github.com/StarRocks/starrocks/pull/52889
where we changed logic in SelectOperator::push_chunk

xhumanoid commented:
@stdpain 

> So the different iteration order of _slot_id_to_index is not a problem.

yes, until you don't have situation like 
1. receive first chunk1
2. iterate by chunk1->_slot_id_to_index and generate one order of `_columns`
3. receive second chunk2
4. iterate by chunk2->_slot_id_to_index and generate another order of `_columns`
5. try to merge chunks in `SelectOperator::pull_chunk`

```
for (size_t i = 0; i < dest_columns.size(); i++) {
   dest_columns[i]->append(*src_columns[i], 0, num_rows);
}
```

it's fully depends from order of columns in chunks, but because iterators produce different order, you have different orders in `_columns` also

xhumanoid commented:
> But this order is not determined by the iteration order of _slot_id_to_index, but by append_column.

yes and no
new logic in push_chunk after

```
Status SelectOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
    if (!_common_exprs.empty()) {

        // Create new chunk
        _curr_chunk = std::make_shared<Chunk>();
        for (const auto& [slot_id, _] : chunk->get_slot_id_to_index_map()) {

           // based on iterator order do _columns.emplace_back(std::move(column));
           // so different operators will have different orders in _curr_chunk based on iterator order
            _curr_chunk->append_column(chunk->get_column_by_slot_id(slot_id), slot_id);
        }
```

stdpain commented:
> > But this order is not determined by the iteration order of _slot_id_to_index, but by append_column.
> 
> yes and no new logic in push_chunk after
> 
> ```
> Status SelectOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
>     if (!_common_exprs.empty()) {
> 
>         // Create new chunk
>         _curr_chunk = std::make_shared<Chunk>();
>         for (const auto& [slot_id, _] : chunk->get_slot_id_to_index_map()) {
> 
>            // based on iterator order do _columns.emplace_back(std::move(column));
>            // so different operators will have different orders in _curr_chunk based on iterator order
>             _curr_chunk->append_column(chunk->get_column_by_slot_id(slot_id), slot_id);
>         }
> ```

it should be the root cause.

stdpain commented:
you can fix it as
```patch
diff --git a/be/src/exec/pipeline/select_operator.cpp b/be/src/exec/pipeline/select_operator.cpp
index 142145c46a2..d64d1f152e0 100644
--- a/be/src/exec/pipeline/select_operator.cpp
+++ b/be/src/exec/pipeline/select_operator.cpp
@@ -87,9 +87,7 @@ bool SelectOperator::need_input() const {
 Status SelectOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
     if (!_common_exprs.empty()) {
         _curr_chunk = std::make_shared<Chunk>();
-        for (const auto& [slot_id, _] : chunk->get_slot_id_to_index_map()) {
-            _curr_chunk->append_column(chunk->get_column_by_slot_id(slot_id), slot_id);
-        }
+        _curr_chunk = chunk;
         SCOPED_TIMER(_conjuncts_timer);
         for (const auto& [slot_id, expr] : _common_exprs) {
             ASSIGN_OR_RETURN(auto col, expr->evaluate(_curr_chunk.get()));
```

xhumanoid commented:
maybe better to validate with @silverbullet233 why we create new chunk

looks like he try to prevent situation when the same chunk can be use somewhere else in parallel 
because we add/remove new expression columns during evaluation

because if we don't have any common expression we do just a copy

```
if (!_common_exprs.empty()) {
   // new chunk
} else {
   _curr_chunk = chunk;
}
```

stdpain commented:
This is a safe operation, see ChunkAccumulateOperator for a similar operation.

xhumanoid committed:
[BugFix] prevent to change column order in SelectOperator::push_chunk

bug introduced in https://github.com/StarRocks/starrocks/pull/52889

iteration order by unordered map is not defined `chunk->get_slot_id_to_index_map()`

so after generate new chunk later in SelectOperator::pull_chunk we can crash

because `_pre_output_chunk` and `_curr_chunk` can have different columns order inside

if types on `append` is matched, final data will be incorrect (merge data from different columns)

if types mismatched, BE will crash

Signed-off-by: Aliaksei Dziomin <diominay@gmail.com>

xhumanoid commented:
@stdpain update pr and version to backport
initial commit was merged in main (3.4 at that moment) and backported into 3.3

so we need to backport into 3.3 and 3.4 only
